### PR TITLE
[rush] Add support for the Yarn package manager

### PR DESCRIPTION
**Pull request by pgonzal**
This PR allows [Yarn](https://yarnpkg.com/) to be used in a Rush repo.  (Hurray!!!)

The symlinking model is the same as Rush's approach for NPM.

### CODE REVIEW

**Comment by pgonzal** [comment]
This PR branch is now pretty much code complete, and ready for review. :-)

File: apps/rush-lib/src/logic/InstallManager.ts [hunk 1]
```diff
@@ -976,6 +994,16 @@ export class InstallManager {
         args.push('--reporter', 'ndjson');
       }
 
+      if (options.networkConcurrency) {
+        args.push('--network-concurrency', options.networkConcurrency.toString());
+      }
+    } else if (this._rushConfiguration.packageManager === 'yarn') {
+      args.push('--ignore-optional');
+      args.push('--link-folder', 'yarn-link');
+      args.push('--cache-folder', 'yarn-cache');
```
Comment (nickpape-msft) [inline apps/rush-lib/src/logic/InstallManager.ts]: config file settings?

File: apps/rush-lib/src/logic/InstallManager.ts [hunk 2]
```diff
@@ -757,6 +761,18 @@ export class InstallManager {
             }
           }
 
+          if (this._rushConfiguration.packageManager === 'yarn') {
+            // Yarn does not correctly detect changes to a tarball, so we need to forcibly clear its cache
+            const yarnRushTempCacheFolder: string = path.join(
+              this._rushConfiguration.commonTempFolder,
+              'yarn-cache', 'v2', 'npm-@rush-temp'
```
Comment (nickpape-msft) [inline apps/rush-lib/src/logic/InstallManager.ts]: >yarn-cache' [](start = 15, length = 11)

magic string

File: apps/rush-lib/src/logic/yarn/YarnShrinkwrapFile.ts [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+
+  private static _decodePackageNameAndSemVer(packageNameAndSemVer: string): IPackageNameAndSemVer {
+    const result: RegExpExecArray | null = YarnShrinkwrapFile.packageNameAndSemVerRegExp.exec(packageNameAndSemVer);
+    if (!result) {
+      // Sanity check -- this should never happen
+      throw new Error('Unable to parse package/semver expression in the Yarn shrinkwrap file: '
+        + JSON.stringify(packageNameAndSemVer));
+    }
+    return {
+      packageName: result[1] || '',
```
Comment (nickpape-msft) [inline apps/rush-lib/src/logic/yarn/YarnShrinkwrapFile.ts]: >|| '' [](start = 29, length = 5)

should these be errors if you aren't able to split it?

File: apps/rush-lib/src/logic/yarn/YarnShrinkwrapFile.ts [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+          // Sanity check -- this should never happen
+          throw new Error('Duplicate @rush-temp package found in the Yarn shrinkwrap file: '
+            + JSON.stringify(key));
+        }
+
+        this._tempProjectNames.push(packageNameAndSemVer.packageName);
+
+        const entry: IYarnShrinkwrapEntry = this._shrinkwrapJson[key];
+
+        // Yarn wrongly performs integrity checking for tarball references.
```
Comment (nickpape-msft) [inline apps/rush-lib/src/logic/yarn/YarnShrinkwrapFile.ts]: >wrongly [](start = 16, length = 7)

how so?

**Comment by pgonzal** [comment]
:shipit:

**Pull request merged**